### ISSUE TITLE
Training on a custom dataset

### ISSUE THREAD

**Issue by FightingKai01** [niki-amini-naieni/CountGD#40]
Hello, sorry to bother you again. Recently, I read your paper and related codes. Although there are still many things I don't understand, I want to try to train the model on my own dataset to further improve the performance of the model, just as you said in the Future Work of your paper.

In terms of dataset production, I referred to the relevant issue and converted coco into odvg format. However, when fsc147_train_odvg.jsonl was converted into fsc147_train_odvg_exemplars.jsonl, there was an extra "exemplars" field. I thought it was three visual examples randomly taken from the "instances" field, but I compared it and found that it was not the case.

I would like to ask how to convert train_odvg.jsonl into train_odvg_exemplars.jsonl. Is there a corresponding script similar to coco2odvg.py? Of course, I hope you can write a more detailed tutorial on making datasets.
Looking forward to your reply! ! !

**Comment by niki-amini-naieni** [comment]
I will work on writing instructions to finetune on one's own dataset. I will update with a timeline soon.

**Comment by FightingKai01** [comment]
So excited, so excited.

Do you have any good suggestions on applying the model to the field of object detection? I am trying hard to try it myself.
I am currently planning to start with the loss function. Of course, my current ability is not enough to change the structure of the model. I am slowly learning the knowledge related to  large model, and now I am just staying at the simple application stage.
 
Your suggestions and reminders will definitely help me a lot.

**Comment by FightingKai01** [comment]
I don't quite understand what you said:This is because the FSC-147 training set does not provide the full bounding box annotations (it only provides the center points of the objects).
The file fsc147_train_odvg_exemplars.jsonl does provide the full bounding box annotations（[x1, y1, x2, y2]）

![Image](https://github.com/user-attachments/assets/87dc703a-028b-4404-bf80-0d50e0d2d505)
After following this suggestion, no other changes need to be made. Can I complete the object detection task training on the FSC-147 training set? My current understanding is that the FSC-147 training set does provide complete bounding boxes.

You can correct me.

**Comment by niki-amini-naieni** [comment]
You are right in that the FSC147 training set does not contain bounding box annotations. However, if your custom dataset **does** have bounding box annotations, then the code for the model can be easily adapted to train on those boxes. Please let me know if the custom dataset you want to train the model on has bounding box annotations, and we can go from there.

**Comment by FightingKai01** [comment]
Maybe there is something wrong with the translation.
I understand that changing the code as suggested in the picture in the previous message can adapt to the object detection training task.

**BUT:**
I think the FSC-147 training set **does** provide complete bounding boxes.The format is [x1, y1, x2, y2].So FSC-147 training set can also be used to train target detection tasks
**As shown**
![Image](https://github.com/user-attachments/assets/23ab9305-e7ba-4de3-b0eb-55f613cb9a73)

**Comment by niki-amini-naieni** [comment]
If you look at the bounding boxes, they all have a height and width of two. The code just takes the center of these coordinates and does not use any further information. The only reason it says "bbox" is to comply with the COCO annotation format. The "bounding boxes" that you have boxed in red are just the center points with a height and width of two. Therefore FSC-147 **does not** provide complete bounding box annotations. The "exemplars" field contains the only bounding box information for FSC-147. 

For example here is one of the training image annotations:
`{"filename": "7.jpg", "height": 384, "width": 384, "detection": {"instances": [{"bbox": [32.744, 295.72639999999996, 34.744, 297.72639999999996], "label": 69, "category": "pepper"}, {"bbox": [22.856, 255.58399999999995, 24.856, 257.58399999999995], "label": 69, "category": "pepper"}, {"bbox": [30.185599999999997, 220.2656, 32.185599999999994, 222.2656], "label": 69, "category": "pepper"}, {"bbox": [44.2064, 178.7216, 46.2064, 180.7216], "label": 69, "category": "pepper"}, {"bbox": [66.60799999999999, 147.3056, 68.60799999999999, 149.3056], "label": 69, "category": "pepper"}, {"bbox": [98.6672, 123.39199999999998, 100.6672, 125.39199999999998], "label": 69, "category": "pepper"}, {"bbox": [140.672, 114.1424, 142.672, 116.1424], "label": 69, "category": "pepper"}, {"bbox": [190.7696, 117.344, 192.7696, 119.344], "label": 69, "category": "pepper"}, {"bbox": [258.25759999999997, 124.61599999999999, 260.25759999999997, 126.61599999999999], "label": 69, "category": "pepper"}, {"bbox": [302.5328, 160.976, 304.5328, 162.976], "label": 69, "category": "pepper"}, {"bbox": [321.152, 205.1984, 323.152, 207.1984], "label": 69, "category": "pepper"}, {"bbox": [326.96959999999996, 258.7232, 328.96959999999996, 260.7232], "label": 69, "category": "pepper"}, {"bbox": [326.96959999999996, 312.2528, 328.96959999999996, 314.2528], "label": 69, "category": "pepper"}]}, "exemplars": [[123, 97, 159, 135], [301, 194, 333, 221], [229, 99, 273, 145]]}`

The first instance is:
`{"bbox": [32.744, 295.72639999999996, 34.744, 297.72639999999996], "label": 69, "category": "pepper"}`

As you said, the bounding box is in the format `[x1, y1, x2, y2]`. Therefore, the width is `(x2-x1)` and the height is `(y2-y1)`, so you get:
`width = 34.744 - 32.744 = 2` and `height = 297.72639999999996 - 295.72639999999996 = 2`.

**Comment by FightingKai01** [comment]
![Image](https://github.com/user-attachments/assets/14976d20-75d0-4061-8f86-4ce9697663ed)
I understand, I found the annotation file of FSC-147 and compared it. As shown in the picture,In order to adapt to the COCO annotation format, the original points are approximated into boxes。

`loss_bbox = F.l1_loss(src_boxes[:, :2], target_boxes[:, :2], reduction='none')`
That is to say, this line of code regresses the coordinates of the upper left corner point and uses the upper left corner point to approximate the center point.

About the preparation of the dataset:
1.Find a dataset containing bounding box annotations and randomly select three as the "examplars" field. This is the key work.

2.At the same time, I also found that the annotation information of each image in the dataset can only contain one category.  **Multiple visual examples + textual hints must refer to the same category. That is, one image can only be trained for one category.** However, different category annotations of the same image can be used to make full use of the image.

Looking forward to the instructions to finetune on one's own dataset.Thanks for your pointers.

**Comment by niki-amini-naieni** [comment]
Hi @FightingKai01, the training code converts the box coordinates from [x1, y1, x2, y2] to [cx, cy, w, h] before the line of code:
`loss_bbox = F.l1_loss(src_boxes[:, :2], target_boxes[:, :2], reduction='none')`
So **it is regressing the center points, not an approximation to them**.  (cx, cy) is the center of the bounding box. If you see [here](https://github.com/niki-amini-naieni/CountGD/blob/be4227a40733a154c80af164a466946b27d289b7/models/GroundingDINO/groundingdino.py#L561), the [cx, cy, w, h] format is converted back to [x1, y1, x2, y2] afterwards.

Regarding the steps you have outlined, for (1) yes, if you want to use the existing training code, you will need exemplars. The method supports training on text only as well, but I will need to make that process clearer. Regarding (2), you could circumvent this issue by including the same image under different names. For example, if image "1.jpg" has annotations for both dogs and cats, you could copy the image under the name "2.jpg" and provide annotations for the cats for "1.jpg" and for dogs in "2.jpg."

**Comment by FightingKai01** [comment]
> **the training code converts the box coordinates from [x1, y1, x2, y2] to [cx, cy, w, h] before the line of code:
loss_bbox = F.l1_loss(src_boxes[:, :2], target_boxes[:, :2], reduction='none')**

After my debugging，I can confirm that **src_boxes** is [cx, cy, w, h], **which can contain the entire target**, which is exactly what I need.It also shows that target_boxes is converted to [cx, cy, **w, h**] **which contains the entire target**.


The annotation of fsc_147_odvg is a bounding box with a fixed width and height of 2, which is approximately a point and meets the coco format. How is this converted into **[cx, cy, w, h] that surrounds the entire target? Here, the w and h of the target are unknown.**

I checked the code before and after you reminded me, and I didn't find the code corresponding to the conversion. I hope you can point out where the code is.

**Comment by niki-amini-naieni** [comment]
The transformation for training is applied [here](https://github.com/niki-amini-naieni/CountGD/blob/321aeba9452ec800e1f711d43dee1d829d3699e5/datasets/odvg.py#L130). The conversion from [x1, y1, x2, y2] to [cx, cy, w, h] happens [here](https://github.com/niki-amini-naieni/CountGD/blob/321aeba9452ec800e1f711d43dee1d829d3699e5/datasets/transforms.py#L296). The code looks like:
`boxes = box_xyxy_to_cxcywh(boxes)`

**Comment by niki-amini-naieni** [comment]
I think that what you are saying is correct. There is some loss in translation. This part is right:
"Similarly, the code becomes like this loss_bbox = F.l1_loss(src_boxes[:, :4], target_boxes[:, :4], reduction='none').At the same time, I have box annotations with Custom Datasets.After training, the inferred bounding box will accurately cover the detection target."

**Comment by FightingKai01** [comment]
Sorry, I'll try changing it again.
The values ​​of all target_boxes are [cx, cy, 2, 2]. Because the loss function only regresses the center point, during inference, the center point of the target box is correct, but the width and height are indeed random.Just like #14.Cannot accurately cover the targets that need to be detected.
On the contrary, I directly changed code to `loss_bbox = F.l1_loss(src_boxes[:, :4], target_boxes[:, :4], reduction='none')`.Use the default dataset for training. The width and length of the target box obtained during inference will be 2.

**Comment by niki-amini-naieni** [comment]
Yes, I believe your understanding is correct.

**Comment by FightingKai01** [comment]
**Regarding the data format of the validation set:** 
The format of annotations.bbox in the coco_val_exemplars.json file is [**x_Top left, y_Top left**, w, h]. It is different from the coordinate format in the training set, but I found that the conversion method used is the same as the training set.[The answer above。](https://github.com/niki-amini-naieni/CountGD/issues/40#issuecomment-2720914586)

I feel like I found the wrong location, what do you think?
[Firstly](https://github.com/niki-amini-naieni/CountGD/blob/cb4e6584906d8b4696bb1aa7730e995cae21fe30/datasets/coco.py#L645) ，[Here](https://github.com/niki-amini-naieni/CountGD/blob/cb4e6584906d8b4696bb1aa7730e995cae21fe30/datasets/coco.py#L474)， then， [there]( https://github.com/niki-amini-naieni/CountGD/blob/cb4e6584906d8b4696bb1aa7730e995cae21fe30/datasets/coco.py#L478)， [Finally](https://github.com/niki-amini-naieni/CountGD/blob/cb4e6584906d8b4696bb1aa7730e995cae21fe30/datasets/transforms.py#L296)

**Comment by niki-amini-naieni** [comment]
Hi, you are correct in that the format of the validation and test annotations is different from the format of the training annotations. However, you missed the following:

Before the transformations are applied, the [x_Top left, y_Top left, w, h] boxes are converted to the [x1, y1, x2, y2] format by the `prepare` function [here](https://github.com/niki-amini-naieni/CountGD/blob/cb4e6584906d8b4696bb1aa7730e995cae21fe30/datasets/coco.py#L375). This function converts the boxes [here](https://github.com/niki-amini-naieni/CountGD/blob/cb4e6584906d8b4696bb1aa7730e995cae21fe30/datasets/coco.py#L426). I have pasted the relevant lines below. As you can see, it sets the width and height of the bounding boxes to w + x_Top_left, y + y_Top_left, resulting in [x_Top_left, y_Top_left, x_Bot_right, y_Bot_right] as desired.

![Image](https://github.com/user-attachments/assets/ef5f4916-8188-4100-80c4-c505c4232860)

**Comment by FightingKai01** [comment]
I made the coco2014 dataset into the dataset format required by CountGD. I want to try to fine-tune the model on this dataset to make it suitable for object detection.



> 1.I found that the value of maxDets in COCO evaluation annotation is limited. [As you mentioned.](https://github.com/niki-amini-naieni/CountGD/issues/22#issuecomment-2562742275)
![Image](https://github.com/user-attachments/assets/3355f6d6-87f4-4f62-82bb-93420a064ebf)
If **num_select** is set to 100, will it affect the application of the model in the field of object detection? Do you have any better suggestions?I prefer to apply it in the field of dense target detection. Similar to [this smart annotation tool](https://trexlabel.com/).

>2.About the early stopping mechanism to get the best model.For object detection, which COCO evaluation indicators should I choose as the basis?For example, select several COCO indicators for weighted combination. I don’t know much about this. Can you tell me your thoughts? Or provide some guidance, I will go deeper into learning.
![Image](https://github.com/user-attachments/assets/7ac2afa1-0c55-4a21-b21f-ccc1ed7eee47)
![Image](https://github.com/user-attachments/assets/efd7c219-5631-4a94-9f22-4f61196c38d2)
![Image](https://github.com/user-attachments/assets/fa2696bb-3d4b-46ac-a5e6-b31e51c41bba)

**Comment by FightingKai01** [comment]
Sorry to bother you again, hope you don't mind, I am indeed a beginner. After days of constant trying, I found the problem and I'm confused.

[training_14.log](https://github.com/user-attachments/files/19416500/training_14.log) is part of my training log on **custom dataset.**Fine-tuning on object detection task using checkpoint_fsc147_best。I found that the AP and AR in the COCO evaluation standard are almost always the following:       Very strange results. The entire bounding box is regressed here.
![Image](https://github.com/user-attachments/assets/d326cd5a-c9ef-4bf1-8bbe-837565a834a8)

MAE and RMSE are also almost unchanged.  Very strange results. The entire bounding box is regressed here.

[training_init.log](https://github.com/user-attachments/files/19416565/training_init.log) is the default training code in your repository without any changes.The changing status of AP ， AR， MAE 和 RMSE is the same as above.

Is there a certain connection? I can't get such a [result](https://github.com/niki-amini-naieni/CountGD/issues/22#issuecomment-2556055402)，although his result is before changing to regressing the entire bounding box.
